Function estrdup() is based on gawk-4.00

Patch by Bryan Mason, https://bugzilla.redhat.com/show_bug.cgi?id=740673

diff -Napur gawk-3.1.7.old/main.c gawk-3.1.7.new/main.c
--- gawk-3.1.7.old/main.c	2009-07-09 12:34:22.000000000 -0700
+++ gawk-3.1.7.new/main.c	2011-09-21 16:25:51.736417000 -0700
@@ -65,6 +65,7 @@ static void nostalgia P((void)) ATTRIBUT
 static void version P((void)) ATTRIBUTE_NORETURN;
 static void init_fds P((void));
 static void init_groupset P((void));
+static char* estrdup(const char *str, size_t len);
 
 /* These nodes store all the special variables AWK uses */
 NODE *ARGC_node, *ARGIND_node, *ARGV_node, *BINMODE_node, *CONVFMT_node;
@@ -701,7 +702,7 @@ add_src(struct src **data, long *num, lo
 	}
 
 	(*data)[*num].stype = stype;
-	(*data)[*num].val = val;
+	(*data)[*num].val = estrdup(val, strlen(val));
 
 #undef INIT_SRC
 }
@@ -1301,6 +1302,18 @@ init_groupset()
 #endif
 }
 
+/* estrdup --- duplicate a string */
+
+char *
+estrdup(const char *str, size_t len)
+{
+  char *s;
+  emalloc(s, char *, len + 1, "estrdup");
+  memcpy(s, str, len);
+  s[len] = '\0';
+  return s;
+}
+ 
 #if defined(HAVE_LOCALE_H)
 
 /* init_locale --- initialize locale info. */
